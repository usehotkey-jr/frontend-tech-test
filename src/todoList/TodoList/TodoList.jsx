// @flow

import React from "react";
import {connect} from "react-redux";
import type {Todo} from "../../newTodo/newTodo.duck";
import {Description, Title} from "./TodoList.styled";
import {selectTodosList} from "./TodosList.selector";
import type {ActionCreator} from "../../_helpers/types/redux";
import {removeTodo} from "../todos.thunk";

type TodoListStateProps = {
    todos: Array<Todo>;
};

type TodoListDispatchProps = {
    removeTodo: ActionCreator<number>;
};

type TodoListProps = TodoListStateProps & TodoListDispatchProps;

export class TodoListComponent extends React.PureComponent<TodoListProps> {
    createRemoveTodo = (id: number) => () => this.props.removeTodo(id);

    render () {
        const todosHtml = this.props.todos.map(todo => (
            <div key={todo.id}>
                <Title>{todo.title}</Title>
                <Description>{todo.description}</Description>
                <span onClick={this.createRemoveTodo(todo.id)}>x</span>
            </div>
        ));

        return (
            <div>
                {todosHtml}
            </div>
        );
    }
}

export const TodoList = connect(
    selectTodosList,
    {
        removeTodo,
    }
)(TodoListComponent);
