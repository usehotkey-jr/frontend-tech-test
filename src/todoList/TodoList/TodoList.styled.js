// @flow

import styled from "styled-components";

export const Title = styled.h3`
  color: brown;
`;

export const Description = styled.h4`
  color: gray;
`;

