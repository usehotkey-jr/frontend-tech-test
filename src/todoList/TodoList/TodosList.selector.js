// @flow

import type {Todo} from "../../newTodo/newTodo.duck";
import {createSelector} from "reselect";
import {selectTodosArray} from "../todos.selector";

export const selectTodosList = createSelector(
    selectTodosArray,
    (todos: Array<Todo>) => ({todos}),
);
