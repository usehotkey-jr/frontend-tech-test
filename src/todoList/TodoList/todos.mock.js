// @flow

import type {Todo} from "../../newTodo/newTodo.duck";

export const todos: Array<Todo> = [
    {
        id: 0,
        title: "Hi",
        description: "Guys",
    },
    {
        id: 1,
        title: "Another",
        description: "Item",
    },
];
