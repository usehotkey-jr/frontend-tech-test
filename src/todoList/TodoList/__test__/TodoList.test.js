import {createTestSnapshot} from "../../../_helpers/jest/createTestSnapshot.ignore";
import {TodoListComponent} from "../TodoList";
import {todosMock} from "../../__mocks__/todos.mock";

describe("TodoList", () => {
    test("should correctly render", () => {
        const props = {
            todos: todosMock,
            removeTodo: jest.fn(),
            updateTodo: jest.fn(),
        };

        const wrapper = createTestSnapshot.all(TodoListComponent, props);

        // Remove button
        wrapper.find("div").forEach(node => node.simulate("click"));
        expect(props.removeTodo.mock.calls).toHaveLength(2);


        const titleAndDescriptions = wrapper.find("input");

        // Change
        titleAndDescriptions.forEach(node => node.simulate("change"));
        expect(props.updateTodo.mock.calls).toHaveLength(4);

        // Blur
        titleAndDescriptions.forEach(node => node.simulate("blur"));
        expect(props.updateTodo.mock.calls).toHaveLength(8);
    });
});
