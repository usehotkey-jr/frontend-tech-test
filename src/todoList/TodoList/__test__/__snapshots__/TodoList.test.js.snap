// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList should correctly render 1`] = `
<TodoListComponent
  todos={
    Array [
      Object {
        "description": "Guys",
        "id": 0,
        "title": "Hi",
      },
      Object {
        "description": "Item",
        "id": 1,
        "title": "Another",
      },
    ]
  }
>
  <div>
    <div
      key="0"
    >
      <styled.h3>
        <h3
          className="sc-bdVaJa fpPKzV"
        >
          Hi
        </h3>
      </styled.h3>
      <styled.h4>
        <h4
          className="sc-bwzfXH dBovyR"
        >
          Guys
        </h4>
      </styled.h4>
      <span
        onClick={[Function]}
      >
        x
      </span>
    </div>
    <div
      key="1"
    >
      <styled.h3>
        <h3
          className="sc-bdVaJa fpPKzV"
        >
          Another
        </h3>
      </styled.h3>
      <styled.h4>
        <h4
          className="sc-bwzfXH dBovyR"
        >
          Item
        </h4>
      </styled.h4>
      <span
        onClick={[Function]}
      >
        x
      </span>
    </div>
  </div>
</TodoListComponent>
`;

exports[`TodoList should correctly render 2`] = `
<div>
  <div>
    <h3
      class="sc-bdVaJa fpPKzV"
    >
      Hi
    </h3>
    <h4
      class="sc-bwzfXH dBovyR"
    >
      Guys
    </h4>
    <span>
      x
    </span>
  </div>
  <div>
    <h3
      class="sc-bdVaJa fpPKzV"
    >
      Another
    </h3>
    <h4
      class="sc-bwzfXH dBovyR"
    >
      Item
    </h4>
    <span>
      x
    </span>
  </div>
</div>
`;

exports[`TodoList should correctly render 3`] = `
<div>
  <div
    key="0"
  >
    <styled.h3>
      Hi
    </styled.h3>
    <styled.h4>
      Guys
    </styled.h4>
    <span
      onClick={[Function]}
    >
      x
    </span>
  </div>
  <div
    key="1"
  >
    <styled.h3>
      Another
    </styled.h3>
    <styled.h4>
      Item
    </styled.h4>
    <span
      onClick={[Function]}
    >
      x
    </span>
  </div>
</div>
`;
